internal/contour/snapshot.go

```diff
@@ -40,7 +52,70 @@ func NewSnapshotHandler(c cache.SnapshotCache, resources []ResourceCache, logger
 	}
 }
 
-// OnChange is called when the DAG is rebuilt
-// and a new snapshot is needed.
+// OnChange is called when the DAG is rebuilt and a new snapshot is needed.
 func (s *SnapshotHandler) OnChange(root *dag.DAG) {
+
+	// Generate new snapshot version.
+	snapshotVersion := s.newSnapshotVersion()
+
+	// Create an snapshot with all xDS resources.
+	snapshot := cache.Snapshot{}
+
+	// Loop through all the xDS resources and pass them to the snapshot
+	for _, r := range s.resources {
+		snapshot.Resources[responseTypeOf(r.TypeURL())] = cache.NewResources(snapshotVersion, asResources(r.Contents()))
+	}
+
+	if err := s.snapshotCache.SetSnapshot(xds.DefaultHash.String(), snapshot); err != nil {
+		s.Errorf("OnChange: Error setting snapshot: %q", err)
+	}
+}
+
+// newSnapshotVersion increments the current snapshotVersion
+// and returns as a string.
+func (s *SnapshotHandler) newSnapshotVersion() string {
+
+	// Reset the snapshotVersion if it ever hits max size.
+	if s.snapshotVersion == math.MaxInt64 {
+		s.snapshotVersion = 0
+	}
+
+	// Increment the snapshot version & return as string.
+	s.snapshotVersion++
+	return strconv.FormatInt(s.snapshotVersion, 10)
+}
+
+// asResources casts the given slice of values (that implement the envoy_xds.Resource
+// interface) to a slice of envoy_xds.Resource. If the length of the slice is 0, it
+// returns nil.
+func asResources(messages interface{}) []envoy_xds.Resource {
+	v := reflect.ValueOf(messages)
+	if v.Len() == 0 {
+		return nil
+	}
+
+	protos := make([]envoy_xds.Resource, v.Len())
+
+	for i := range protos {
+		protos[i] = v.Index(i).Interface().(envoy_xds.Resource)
+	}
+
+	return protos
+}
+
+// responseTypeOf converts a typeURL to an envoy_xds.ResponseType.
+func responseTypeOf(typeURL string) envoy_xds.ResponseType {
```

Comment:
it'd probably be slightly more efficient to use a map from `typeURL` to `ResponseType` here, rather than having to execute the switch every time. May not make any material difference in practice, though.

Reply:
Yeah I'll do that, it was the initial thought to make Resources a map from the start, but I'll do that work here.